### ISSUE TITLE
Class imbalance for RPN

### ISSUE THREAD

**Issue by MBleeker** [matterport/Mask_RCNN#43]
Hi,

For my master thesis I am working on  the topic of text detection in images and video frames. I implemented a modified version of Faster R-CNN/Mask R-CNN, which is very similar to your implementation, but tailored for text detection. 

On average there are ~54 positive text anchors in the images I use, which result in a mini batch of ~54 positive and ~200 negative examples per image ( I use mini batches of size 254 per image). The problem I encounter is that my network overfits on only negative prediction (because on average, there are many more negative than positive examples). Now  a few simple solutions would be to 1) take a smaller mini batch size(for example 112), 2) to remove all the images in which there are not enough positive examples from the data set or 3) use a weighted loss function. 

I inspected your code very carefully, but (as far as I can see) in your implementation this doesn't seem to be an issue. Was this positive/negative class imbalance also a problem in your implementation, and if so, how did you solve this problem?

Thanks!

Maurits

**Comment by MBleeker** [comment]
Using balanced mini batches solved the problem!

**Comment by MBleeker** [comment]
Thanks for your answer.

That is an interesting approach. Does it work well? I mean, the gradients you are using when updating the weights will be based on a different amount of samples every epoch. Does this work well with a fixed learning rate. When only having a few examples in your batch, the gradients will be very noisy right?

To prevent overfitting on negative examples, I used a weighted loss function. I scale up the loss of the positive examples with the ratio = (batch size/ 2)/ (positive examples), the same for the negative examples. By using this I can keep the batch size constant, and it will be impossible for the network to learn to overfit the negative examples.

I still do not understand why they use a 1:3 positive to negative ratio in the paper. I mean, it will still be really easy to overfit on only predicting negative examples, which will result in a correct classification of 2:3, which is better than random 1:2. 

Maurits

**Comment by waleedka** [comment]
> When only having a few examples in your batch, the gradients will be very noisy right?

You're right, but most images would have around 40 or so positive ROIs (and therefore, 80 negative ROIs). Even an image with one object would have a total of 20 or more ROIs. 

> I still do not understand why they use a 1:3 positive to negative ratio in the paper.

My guess is that they experimented with different ratios and found that this worked the best. It would be interesting to compare it to your weighted loss approach and see if you get better results.

**Comment by Amrimn** [comment]
> Thanks for your answer.
> 
> That is an interesting approach. Does it work well? I mean, the gradients you are using when updating the weights will be based on a different amount of samples every epoch. Does this work well with a fixed learning rate. When only having a few examples in your batch, the gradients will be very noisy right?
> 
> To prevent overfitting on negative examples, I used a weighted loss function. I scale up the loss of the positive examples with the ratio = (batch size/ 2)/ (positive examples), the same for the negative examples. By using this I can keep the batch size constant, and it will be impossible for the network to learn to overfit the negative examples.
> 
> I still do not understand why they use a 1:3 positive to negative ratio in the paper. I mean, it will still be really easy to overfit on only predicting negative examples, which will result in a correct classification of 2:3, which is better than random 1:2.
> 
> Maurits

Hello @MBleeker ,
I'm facing the same problem of an imbalanced class can you please show me how you used a weighted loss function?
I am suffering from overfitting.
thank you in advance
